File: scripts/science-equations/publish-internet-archive.mjs
#!/usr/bin/env node

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { spawnSync } from "node:child_process";

function arg(name) {
  const i = process.argv.indexOf(name);
  return i >= 0 ? process.argv[i + 1] : null;
}

function sha256File(file) {
  const hash = crypto.createHash("sha256");
  hash.update(fs.readFileSync(file));
  return hash.digest("hex");
}

function md5File(file) {
  const hash = crypto.createHash("md5");
  hash.update(fs.readFileSync(file));
  return hash.digest("hex");
}

function verifyLocalManifest(manifestPath, manifest) {
  const dir = path.dirname(manifestPath);
  const files = [];
  for (const shard of manifest.shards ?? []) {
    const file = path.resolve(dir, shard.file);
    if (!fs.existsSync(file)) throw new Error(`missing shard: ${shard.file}`);
    const bytes = fs.statSync(file).size;
    const sha256 = sha256File(file);
    const md5 = md5File(file);
    if (bytes !== shard.bytes) throw new Error(`${shard.file}: ${bytes} bytes != manifest ${shard.bytes}`);
    if (sha256 !== shard.sha256) throw new Error(`${shard.file}: SHA-256 mismatch`);
    if (shard.md5 && md5 !== shard.md5) throw new Error(`${shard.file}: MD5 mismatch`);
    files.push({ file, name: shard.file, bytes, sha256, md5 });
  }
  if (files.length === 0) throw new Error("manifest has no Parquet shards");
  return files;
}

async function fetchMetadata(identifier) {
  const response = await fetch(`https://archive.org/metadata/${encodeURIComponent(identifier)}`);
  if (!response.ok) throw new Error(`Internet Archive metadata HTTP ${response.status}`);
  return response.json();
}

async function waitForRemoteFiles(identifier, expected, attempts = 18) {
  let last = null;
  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try {
      const metadata = await fetchMetadata(identifier);
      const remote = new Map((metadata.files ?? []).map((file) => [file.name, file]));
      const missing = [];
      for (const local of expected) {
        const found = remote.get(local.name);
        if (!found) {
          missing.push(`${local.name}:missing`);
          continue;
        }
        if (Number(found.size) !== local.bytes) missing.push(`${local.name}:bytes`);
        if (found.md5 && found.md5 !== local.md5) missing.push(`${local.name}:md5`);
      }
      if (missing.length === 0) return metadata;
      last = missing.join(", ");
    } catch (error) {
      last = error instanceof Error ? error.message : String(error);
    }
    if (attempt < attempts) {
      await new Promise((resolve) => setTimeout(resolve, Math.min(attempt * 5000, 30000)));
    }
  }
  throw new Error(`Internet Archive verification did not converge: ${last}`);
}

async function main() {
  const manifestArg = arg("--manifest");
  const title = arg("--title");
  const description = arg("--description") ?? "Scientific Equation Atlas source snapshot";
  const creator = arg("--creator") ?? "Scientific Equation Atlas";
  const iaCli = process.env.IA_CLI || "ia";
  if (!manifestArg) throw new Error("--manifest is required");
  if (!process.env.IA_ACCESS_KEY_ID || !process.env.IA_SECRET_ACCESS_KEY) {
    throw new Error("IA_ACCESS_KEY_ID and IA_SECRET_ACCESS_KEY are required in the executor environment");
  }

  const manifestPath = path.resolve(manifestArg);
  const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
  if (manifest.format !== "parquet") throw new Error("only Parquet lake manifests may be published");
  if (manifest.stage !== "extracted" && manifest.stage !== "normalized" && manifest.stage !== "deduplicated") {
    throw new Error(`unsupported lake stage: ${manifest.stage}`);
  }
  const identifier = manifest.internet_archive?.identifier;
  if (!identifier) throw new Error("manifest.internet_archive.identifier is required");

  const localFiles = verifyLocalManifest(manifestPath, manifest);
  const uploadFiles = [...localFiles.map((entry) => entry.file), manifestPath];
  const metadata = [
    "mediatype:data",
    `title:${title ?? `Scientific Equation Atlas — ${manifest.source_id} — ${manifest.source_snapshot}`}`,
    `creator:${creator}`,
    `description:${description}`,
    `subject:Scientific Equation Atlas;equations;Parquet;open data;${manifest.source_id}`,
  ];

  const command = ["upload", identifier, ...uploadFiles, "--retries", "10"];
  for (const value of metadata) command.push("--metadata", value);
  const run = spawnSync(iaCli, command, { encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] });
  if (run.status !== 0) throw new Error(run.stderr || run.stdout || "Internet Archive upload failed");

  const expected = [
    ...localFiles,
    {
      file: manifestPath,
      name: path.basename(manifestPath),
      bytes: fs.statSync(manifestPath).size,
      sha256: sha256File(manifestPath),
      md5: md5File(manifestPath),
    },
  ];
  await waitForRemoteFiles(identifier, expected);

  const publication = {
    identifier,
    metadata_url: `https://archive.org/metadata/${identifier}`,
    details_url: `https://archive.org/details/${identifier}`,
    files: expected.map(({ name, bytes, sha256, md5 }) => ({ name, bytes, sha256, md5 })),
    verified: true,
  };
  process.stdout.write(`${JSON.stringify(publication, null, 2)}\n`);
}

main().catch((error) => {
  process.stderr.write(`${error instanceof Error ? error.stack : String(error)}\n`);
  process.exitCode = 1;
});

File: scripts/science-equations/publish-internet-archive.py
#!/usr/bin/env python3
"""Publish a verified Parquet dataset item to Internet Archive without silent overwrite."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import re
import subprocess
import time
import urllib.request
from pathlib import Path


def digest_file(path: Path, algorithm: str) -> str:
    digest = hashlib.new(algorithm)
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def sanitize(value: str) -> str:
    return re.sub(r"[^A-Za-z0-9._-]+", "-", value).strip("-").lower()


def metadata(identifier: str) -> dict:
    with urllib.request.urlopen(f"https://archive.org/metadata/{identifier}", timeout=60) as response:
        return json.load(response)


def remote_files(meta: dict) -> dict[str, dict]:
    return {entry.get("name"): entry for entry in meta.get("files", []) if entry.get("name")}


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--manifest", required=True, type=Path)
    parser.add_argument("--identifier")
    parser.add_argument("--ia-bin", default="ia")
    parser.add_argument("--title")
    parser.add_argument("--description", default="Scientific Equation Atlas Parquet snapshot")
    parser.add_argument("--verify-attempts", type=int, default=18)
    args = parser.parse_args()

    if not os.environ.get("IA_ACCESS_KEY_ID") or not os.environ.get("IA_SECRET_ACCESS_KEY"):
        raise SystemExit("IA_ACCESS_KEY_ID and IA_SECRET_ACCESS_KEY are required in the executor environment")

    manifest = json.loads(args.manifest.read_text(encoding="utf-8"))
    source_id = manifest["source_id"]
    snapshot = manifest["source_snapshot"]
    stage = manifest["stage"]
    if not args.identifier:
        clean_source = sanitize(source_id)
        hex_match = re.search(r"[0-9a-fA-F]{16,64}", snapshot)
        clean_snap = hex_match.group(0)[:16] if hex_match else sanitize(snapshot)[:16]
        clean_stage = sanitize(stage)
        raw_id = f"atlas-{clean_source}-{clean_snap}-{clean_stage}"
        if len(raw_id) > 80:
            raw_id = f"atlas-{clean_source[:30]}-{clean_snap[:16]}-{clean_stage[:10]}"
        identifier = raw_id
    else:
        identifier = args.identifier

    if len(identifier) > 80:
        raise SystemExit(f"Internet Archive identifier must be <= 80 chars, got {len(identifier)}: {identifier}")
    base = args.manifest.parent
    paths = [base / shard["file"] for shard in manifest["shards"]] + [args.manifest]
    expected = {
        path.name: {"size": path.stat().st_size, "md5": digest_file(path, "md5"), "sha256": digest_file(path, "sha256")}
        for path in paths
    }

    try:
        current = metadata(identifier)
    except Exception:
        current = {}
    existing = remote_files(current)
    for name, local in expected.items():
        if name not in existing:
            continue
        remote = existing[name]
        remote_size = int(remote.get("size", -1)) if str(remote.get("size", "")).isdigit() else -1
        remote_md5 = remote.get("md5")
        if remote_size != local["size"] or (remote_md5 and remote_md5 != local["md5"]):
            raise SystemExit(f"refusing silent overwrite: {identifier}/{name} already exists with different bytes")

    to_upload = [str(path) for path in paths if path.name not in existing]
    if to_upload:
        title = args.title or f"Scientific Equation Atlas — {source_id} — {snapshot} — {stage}"
        command = [
            args.ia_bin,
            "upload",
            identifier,
            *to_upload,
            "--retries",
            "10",
            "--metadata",
            "mediatype:data",
            "--metadata",
            f"title:{title}",
            "--metadata",
            "creator:Scientific Equation Atlas",
            "--metadata",
            f"description:{args.description}",
        ]
        subprocess.run(command, check=True)

    for attempt in range(1, args.verify_attempts + 1):
        files = remote_files(metadata(identifier))
        failures = []
        for name, local in expected.items():
            remote = files.get(name)
            if not remote:
                failures.append(f"missing:{name}")
                continue
            remote_size = int(remote.get("size", -1)) if str(remote.get("size", "")).isdigit() else -1
            if remote_size != local["size"]:
                failures.append(f"size:{name}")
            if remote.get("md5") and remote["md5"] != local["md5"]:
                failures.append(f"md5:{name}")
        if not failures:
            publication = {
                "identifier": identifier,
                "metadata_url": f"https://archive.org/metadata/{identifier}",
                "details_url": f"https://archive.org/details/{identifier}",
                "files": expected,
                "verified": True,
            }
            print(json.dumps(publication, indent=2, sort_keys=True))
            return 0
        if attempt < args.verify_attempts:
            time.sleep(min(attempt * 5, 20))

    raise SystemExit(f"Internet Archive verification did not converge for {identifier}")


if __name__ == "__main__":
    raise SystemExit(main())
